scrapscript.py

```diff
@@ -466,125 +530,132 @@ def gensym_reset() -> None:
-        if base == 85:
-            l = Bytes(base64.b85decode(token.value))
-        elif base == 64:
-            l = Bytes(base64.b64decode(token.value))
-        elif base == 32:
-            l = Bytes(base64.b32decode(token.value))
-        elif base == 16:
-            l = Bytes(base64.b16decode(token.value))
-        else:
-            raise ParseError(f"unexpected base {base!r} in {token!r}")
-        return l
-    elif isinstance(token, StringLit):
-        return String(token.value)
-    elif token == Operator("..."):
-        if tokens and isinstance(tokens[0], Name):
-            name = tokens[0].value
-            tokens.pop(0)
-            return Spread(name)
-        else:
-            return Spread()
-    elif token == Operator("|"):
-        expr = parse_binary(tokens, PS["|"].pr)  # TODO: make this work for larger arities
-        if not isinstance(expr, Function):
-            raise ParseError(f"expected function in match expression {expr!r}")
-        cases = [MatchCase(expr.arg, expr.body)]
-        while tokens and tokens[0] == Operator("|"):
-            tokens.pop(0)
+def parse_unary(tokens: Peekable, p: float) -> "Object":
+    try:
+        token = next(tokens)
+        l: Object
+        if isinstance(token, IntLit):
+            return Int(token.value)
+        elif isinstance(token, FloatLit):
+            return Float(token.value)
+        elif isinstance(token, Name):
+            # TODO: Handle kebab case vars
+            return Var(token.value)
+        elif isinstance(token, Hash):
+            if isinstance(variant := tokens.peek(), Name):
+                next(tokens)
+                # It needs to be higher than the precedence of the -> operator so that
+                # we can match variants in MatchFunction
+                # It needs to be higher than the precedence of the && operator so that
+                # we can use #true() and #false() in boolean expressions
+                # It needs to be higher than the precedence of juxtaposition so that
+                # f #true() #false() is parsed as f(TRUE)(FALSE)
+                return Variant(variant.value, parse_binary(tokens, PS[""].pr + 1))
+            else:
+                raise UnexpectedTokenError(variant)
+        elif isinstance(token, BytesLit):
+            base = token.base
+            if base == 85:
+                l = Bytes(base64.b85decode(token.value))
+            elif base == 64:
+                l = Bytes(base64.b64decode(token.value))
+            elif base == 32:
+                l = Bytes(base64.b32decode(token.value))
+            elif base == 16:
+                l = Bytes(base64.b16decode(token.value))
+            else:
+                raise ParseError(f"unexpected base {base!r} in {token!r}")
+            return l
+        elif isinstance(token, StringLit):
+            return String(token.value)
+        elif token == Operator("..."):
+            try:
+                if isinstance(tokens.peek(), Name):
```

Comment:
This kind of stuff is really subtle and I am wondering if there is a way we can clean it up somehow. Maybe some kind of `have_tokens()` function? :/